### DISCUSSION TITLE
LoadBalancer similar to ReactPHP-based "PHP-PM"

### DISCUSSION THREAD

**Discussion by fakharak** [swoole/swoole-src#4266] [Q&A]
PHP-PM claims to have built-in LoadBalancer.
See:
https://github.com/php-pm/php-pm

Can we make something similar with Swoole ?

**Comment by linuxd3v** [comment]
There is already in a load balancing built-in. 
You can configure swoole to have any number of worker processes through workers_num - and use dispatch_mode for various load -balancing modes (see SWOOLE_PROCESS docs). 
Swoole restarts worker process if it goes down.

**Comment by ClosetGeek-Git** [comment]
PHP-PM spawns new PHP processes using userspace proc_open, and then communicates with them over php_streams in userspace. This means that if a spawned process needs to be restarted that it's whole state needs to be reinitialized, not just PHP code, but the whole PHP VM as well as any persistent resources.
Swoole on the other hand uses fork and then communicates with the child processes using swoole's native reactor and coroutine sockets, it uses a pre-fork pattern minimizing time to restart children, and is able to maintain persistent resources across child restarts.

**Reply by ClosetGeek-Git** [reply]
Swoole is also able to restart PHP within children without the entire child process having to be restarted. In this way swoole is not just able to create and manage processes, but acts more like an SAPI. I think it's safe to call Swoole a hybrid SAPI where PHP-PM is just a process manager.
